jgrapht-core/src/main/java/org/jgrapht/traverse/DepthFirstIterator.java

```diff
@@ -34,42 +34,14 @@
  * @param <V> the graph vertex type
  * @param <E> the graph edge type
  *
- * @author Liviu Rau
- * @author Barak Naveh
- * @since Jul 29, 2003
+ * @author Joris Kinable
  */
 public class DepthFirstIterator<V, E>
-    extends CrossComponentIterator<V, E, DepthFirstIterator.VisitColor>
+    extends CrossComponentIterator<V, E, DepthFirstIterator.SearchNodeData>
 {
-    /**
-     * Sentinel object. Unfortunately, we can't use null, because ArrayDeque won't accept those. And
-     * we don't want to rely on the caller to provide a sentinel object for us. So we have to play
-     * typecasting games.
-     */
-    public static final Object SENTINEL = new Object();
-
-    /**
-     * Standard vertex visit state enumeration.
-     */
-    protected static enum VisitColor
-    {
-        /**
-         * Vertex has not been returned via iterator yet.
-         */
-        WHITE,
-
-        /**
-         * Vertex has been returned via iterator, but we're not done with all of its out-edges yet.
-         */
-        GRAY,
-
-        /**
-         * Vertex has been returned via iterator, and we're done with all of its out-edges.
-         */
-        BLACK
-    }
 
-    private Deque<Object> stack = new ArrayDeque<>();
+    //private Stack<V> stack = new Stack<>();
+    private Stack<StackEntry> stack = new Stack<>();
```

Comment:
Hi @jkinable. I've not had a look at the rest of the PR yet, but I wanted to say, looking at this, that I feel a bit iffy with `Stack` being used here rather than `Deque`.

If I may, I would like to suggest using `Deque` here instead, for a couple of reasons.

Firstly, according to the `ArrayDeque` [javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayDeque.html), it's explained that ArrayDeque "is likely to be faster than Stack when used as a stack".

And finally, the `Stack` [javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/Stack.html) explains that it extends `Vector`, which IIRC has built-in synchronization, which I would reasonably guess makes `Stack` slower than necessary in this case, since AFAICT the stack isn't being used in an asynchronous fashion.